File: 1to150/62_uniquePath.java
/*
1. state function: dp[m][n], each dp[i][j] represents how many different ways a robot could reach here

2. init state function： dp[0][n] = 1 and dp[m][0] = 1 因为对于第一row，机器人只能从左边走过来 对于第一列机器人只能从上面走下来

3.transfer function: 对于 i j项， 假设第二行 第三列， 机器人想要走过来势必要么是从上走下来，或者从左走过来，那么对于这一个的位置的走法，就是 dp[i][j - 1] + dp[i - 1][j]， 相当于有多少种走到上面那一格

4.result, dp[m-1][n-1] 最后一位的总走法
*/
class Solution {
    public int uniquePaths(int m, int n) {
        /* each grid represents how many ways it could get to this point*/
        int[][] dp = new int[m][n];
        
        //init 
        for (int i = 0; i < m; i++){
            dp[i][0] = 1;
        }
        for (int j = 0; j < n; j++){
            dp[0][j] = 1;
        }
        
        //transfer function,注意这里都是从第二项开始走，因为第一项都已init过了
        for (int i = 1; i < m; i ++){
            for (int j = 1; j < n; j++){
                dp[i][j] = dp[i-1][j] + dp[i][j - 1];
            }
        }
        
        return dp[m-1][n-1];
    }
}

File: 1to150/62_uniquePath.py
'''
1. state: 2d array, 每一个位置代表走到这个grid能走的总走法
2. init： 第一row机器人只能走左走到右，所以每一个grid都是1
          同理，第一列的，机器人只能从上走到下面， 所以也是1
          
3. transfer：
   1 1 1 1 1 
   1   X
   1 X ？       想要走到这地方，机器人可以从上走下来，也可从左走来dp[i - 1][j] + dp[i][j - 1] 可见我们需要的是过去的信息
   1            topDown
 
4. return 最后的dp得到走到这个位置的所有走法
''' 

class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        dp = [[0 for j in range(n)] for i in range(m)]    
        
        #init
        for i in range(m):
            dp[i][0] = 1
            
        for j in range(n):
            dp[0][j] = 1
        
            
        #transfer
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
                
        return dp[m - 1][n - 1]